packages/cardano-services/src/DbSyncProvider.ts

```diff
@@ -3,19 +3,113 @@ import { Pool } from 'pg';
 
 const HEALTH_CHECK_QUERY = 'SELECT 1';
 
-export abstract class DbSyncProvider implements Provider {
-  public db: Pool;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type AnyArgs = any[];
```

Comment:
In general we're following a pattern of 2 args: `(props, dependencies)` so it could probably be typed with generic arguments <TProps, TDependencies>

Reply:
You hit two distinct topics here, @mkazlauskas .

`type AnyArgs = any[]` is more or less mandatory as TypeScript requires the `constructor` of a _mixin_ to be of type `new(...args: any[]) => T`; but this is just an implementation detail. We can (and IMHO we should) achieve what you are describing working on the tuples inferred from the arguments of the two constructors.

While about the core of the suggestion, I 100% agree we should do that. We just need to simplify the way to achieve it, please let me explain in details.

Now the code simply concatenates the two tuples inferred from the constructors: two arbitrary tuples. trying to merge the types we have as elements of arbitrary tuples risk to increase a lot the complexity of the code. As an example, the `RunnableModule.constructor` itself doesn't respect the `(props, deps)` pattern.
I thought that in order to achieve this more standardization would be helpful to standardize all or `Provider` implementations (and probably `Service`s as well) to something similar to:

```typescript
type ProviderConfiguration = { [k: string]?: unknown };

interface ProviderDependencies {
  logger: Logger;
  db?: Pool;
  cache?: MemCache;
  ...
}

interface ProviderImplementation {
  new(config: ProviderConfiguration, dependencies: ProviderDependencies): Provider;
}
```

This will make the code required much more easy and clean than the one required now.
On the other side to implement this on all our `Provider`s and `Service`s (which means do the same for all the configuration and dependencies interfaces) would require a dedicated epic split in several tasks.

What do you think about?

Reply:
> working on the tuples inferred from the arguments of the two constructors.

> trying to merge the types we have as elements of arbitrary tuples risk to increase a lot the complexity of the code

Sounds good, but if there's no obvious/easy way to do that then I'm fine with any[] as long as it's contained within a single class (any[] doesn't bleed to other classes).

> I thought that in order to achieve this more standardization would be helpful to standardize all or Provider implementations

Agree - we've been using a similar pattern for the most part. I'm sure there are some different signatures, but it should be an easy task to refactor to make all services compliant. Note: I think this is more of a "Service" thing than a "Provider" thing - provider is just an interface, where services within `cardano-services` package implements it.

What are your thoughts on this?:

> I'm wondering if it would make sense to use "DbSyncProvider" as a mixin (the other way around) and have all the mixin logic within RunnableModule

Reply:
> Note: I think this is more of a "Service" thing than a "Provider" thing - provider is just an interface, where services within cardano-services package implements it.

I agree this is not a `Provider` related topic, (sorry if I was not clear enough); in the example on the fly I called it `ProviderImplementation` but I was referring strictly to the domain of `cardano.services` domain.

> I'm wondering if it would make sense to use "DbSyncProvider" as a mixin (the other way around) and have all the mixin logic within RunnableModule

I'm not sure I got the point; if you are proposing to leave `DbSyncProvider` a standard class and implementing the `RunnableModule` as the mixin; this is not a viable solution due to a limitation of my implementation: it doesn't let us the overload the methods of the mixin, and the design of `RunnableModule` requires subclasses do that.

If I answered to another question than your one, please help me better understanding.

Reply:
Yes, I thought maybe DbSyncProvider can be a mixin, so maybe it could look like `extends HttpService(RunnableModule, DbSyncProvider)`

Reply:
As agreed in the standup, I'm resolving this